org.osgi.test.assertj/src/main/java/org/osgi/test/assertj/bundle/AbstractBundleAssert.java

```diff
@@ -0,0 +1,366 @@
+			failWithMessage(
+				"%nExpecting%n  <%s>%nto have been modified before or at:%n  <%d>%n but was modified at:%n  <%d>",
+				actual, expectedTime, actual.getLastModified());
+		}
+		return myself;
+	}
+
+	private static long getTime(Object expected) {
+		if (expected == null) {
+			throw new IllegalArgumentException("expected cannot be null");
+		} else if (expected instanceof Long) {
+			return (Long) expected;
+		} else if (expected instanceof Date) {
+			return ((Date) expected).getTime();
+		} else if (expected instanceof Instant) {
+			return ((Instant) expected).toEpochMilli();
+		} else {
+			throw new IllegalArgumentException("Expected must be a long, Date or Instant");
+		}
+	}
+
+	public SELF isInState(int expected) {
+		isNotNull();
+		if (actual.getState() != expected) {
+			failWithMessage("%nExpecting%n  <%s>%nto be in state:%n  <%s>%n but was in state:%n  <%s>", actual,
+				stateToString(expected), stateToString(actual.getState()));
+		}
+		return myself;
+	}
+
+	public SELF isNotInState(int expected) {
+		isNotNull();
+		final String expectedState = stateToString(expected);
+		if (actual.getState() == expected) {
+			failWithMessage("%nExpecting%n  <%s>%nnot to be in state:%n  <%s>%n but it was", actual,
+				expectedState);
+		}
+		return myself;
+	}
+
+	public SELF isInStateMask(int mask) {
+		isNotNull();
+		if (mask <= 0 || mask >= ACTIVE * 2) {
+			throw new IllegalArgumentException("Mask testing for an illegal state: " + mask);
+		}
+		if ((actual.getState() & mask) == 0) {
+			final String states = IntStream.of(STATES)
+				.filter(x -> (x & mask) != 0)
+				.mapToObj(AbstractBundleAssert::stateToString)
+				.collect(joining(","));
+			failWithMessage("%nExpecting%n  <%s>%nto be in one of states:%n  [%s]%n but was in state:%n  <%s>",
+				actual,
+				states, stateToString(actual.getState()));
+		}
+		return myself;
+	}
+
+	public SELF isNotInStateMask(int mask) {
+		isNotNull();
+		if (mask <= 0 || mask >= ACTIVE * 2) {
+			throw new IllegalArgumentException("Mask testing for an illegal state: " + mask);
+		}
+		if ((actual.getState() & mask) != 0) {
+			final String states = IntStream.of(STATES)
+				.filter(x -> (x & mask) != 0)
+				.mapToObj(AbstractBundleAssert::stateToString)
+				.collect(joining(","));
+			failWithMessage("%nExpecting%n  <%s>%nto not be in one of states:%n  [%s]%n but was in state:%n  <%s>",
+				actual,
+				states, stateToString(actual.getState()));
+		}
+		return myself;
+	}
+
+	private static final int[] STATES = {
+		UNINSTALLED, INSTALLED, RESOLVED, STARTING, STOPPING, ACTIVE
+	};
+
+	public SELF isInOneOfStates(int... states) {
+		return isInStateMask(IntStream.of(states)
+			.filter(x -> stateToString(x) != null) // Will throw IAE if
+													// necessary
+			.reduce(0, (x, y) -> x | y));
+	}
+
+	public SELF isNotInOneOfStates(int... states) {
+		return isNotInStateMask(IntStream.of(states)
+			.filter(x -> stateToString(x) != null) // Will throw IAE if
```

Comment:
If you need to see of the supplied state value is valid, just AND it with the OR of all the valid state values. Must faster and simpler.

Reply:
Again, the idea was that each of the states would be a single bit and that if they weren't it would throw IAE (the call to `stateToString()` was a semi-hack to trigger the IAE). If you wanted to allow test masks, you can use `isInStateMask()`.

Reply:
Still, there is no need for strings or switches here. As I mention above, you can test that each state arg has exactly one bit set and error if not.

Reply:
The main idea of the string (and the switch that generates it) is to produce a human-friendly error message in the event of a failing assertion (naming the state(s) instead of printing a number).

I'll try and simplify it.

Reply:
Done.